Add theme cache to the inspector

KoBeWi opened the pull request:
Adds theme cache to everything in `editor_inspector.cpp`, i.e. EditorInspector, EditorProperty etc.
Unlike other theme caches, editor properties and other sub-classes don't have their own cache. Instead they use a theme cache from a singleton inspector. This should make inspector updates faster, while the editor startup might be slower (due to different theme update method and a new inspector, which exists only to hold theme cache). I did not do any benchmarks, but everything seems to work and appear as it did before, at least.

AThousandShips reviewed (approved):
LGTM

KoBeWi commented:
Rebased after #93435
I also realized I didn't put theme icons in the cache. There is so many of them though 🤔

passivestar commented:
Works except it changed the size of categories and sections and also section arrows are pointing in the wrong direction 🙃

| Master | PR |
|--------|--------|
| ![master](https://github.com/godotengine/godot/assets/60579014/7b62aec0-e18f-45c4-9ed7-5c9fc9800efc) | ![pr](https://github.com/godotengine/godot/assets/60579014/328c802f-c9a3-4436-82b6-369ca1bc9c6f) |

passivestar commented:
Works as expected after the last changes

akien-mga commented:
Needs a rebase, and this seemed to crash when generating Mono glue on CI, but it's not clear whether it's from this PR or just bad timing.

akien-mga commented:
Indeed C# glue generation still seems to crash. We don't have debug symbols on the CI build to know why though, would need to be tested locally.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -2664,6 +2678,17 @@ Button *EditorInspector::create_inspector_action_button(const String &p_text) {
+	if (!theme_singleton) {
+		theme_singleton = memnew(EditorInspector);
+		theme_singleton->hide();
+		if (EditorNode::get_singleton()) {
```
This fixed the crash, but technically causes a leak.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -2664,6 +2678,17 @@ Button *EditorInspector::create_inspector_action_button(const String &p_text) {
+	if (!theme_singleton) {
+		theme_singleton = memnew(EditorInspector);
+		theme_singleton->hide();
+		if (EditorNode::get_singleton()) {
+			EditorNode::get_singleton()->get_gui_base()->add_child(theme_singleton);
+		}
```
This condition fixed the crash, but technically causes a leak.

akien-mga replied:
Can we add it as child somewhere else as fallback to make sure it will get freed?

I have to say it feels a bit hacky to keep a static shadow Inspector for theme caching, isn't that pretty wasteful in number of objects and memory used to basically store a config file?

KoBeWi replied:
It's still better than storing the cache in each inspector property individually. Unfortunately the main inspector is initialized too late, so it couldn't be used. Also, the fake inspector is empty, so it doesn't take that much.

> Can we add it as child somewhere else as fallback to make sure it will get freed?

SceneTree does not exist at all in this situation, so no.

akien-mga replied:
So what should we do? I'm not too fond of knowingly adding a leak. It's not for the main usage of the editor so it's not too bad but it would still be better without introducing a technical "bug".

KoBeWi replied:
I used a deferred static method.

Calinou reviewed (approved):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/0a4aedb36065f66fc7e99cb2e6de3e55242f9dfb), it works as expected.

I can't really notice a difference in terms of startup time or inspector responsiveness though (tested on a Linux x86_64 editor build with `optimize=speed lto=full`).

The startup + shutdown time difference across 100 runs is very small:

```
Benchmark 1: bin/godot.linuxbsd.editor.x86_64 /tmp/4/project.godot --quit     <---- This PR
  Time (mean ± σ):      3.799 s ±  0.182 s    [User: 2.135 s, System: 0.612 s]
  Range (min … max):    2.878 s …  4.048 s    100 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Benchmark 2: bin/godot.linuxbsd.editor.x86_64.master /tmp/4/project.godot --quit
  Time (mean ± σ):      3.814 s ±  0.198 s    [User: 2.133 s, System: 0.612 s]
  Range (min … max):    2.848 s …  4.096 s    100 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Summary
  bin/godot.linuxbsd.editor.x86_64 /tmp/4/project.godot --quit ran
    1.00 ± 0.07 times faster than bin/godot.linuxbsd.editor.x86_64.master /tmp/4/project.godot --quit
```

The stripped binary with this PR is 4 KB smaller.

akien-mga commented:
Needs rebase.

akien-mga commented:
> Needs rebase.

Bump, should be fine to merge now.

KoBeWi commented:
I ran into some problems last time I tried to rebase and then realized that the implementation of `theme_singleton` is rather questionable. I'll try to rework it into something that makes more sense.

KoBeWi committed:
Add theme cache to the inspector

KoBeWi commented:
Ok I reworked the PR.

The main problem with the previous code was that it didn't work if the nodes were created outside inspector. I moved the theme cache to the nodes, however the inspector still holds its own copy for inspector controls. If a control (category, section) is created under inspector, it will copy its cache, so rebuilding the tree will avoid theme lookups. If the control is standalone, it will initialize its own cache. I achieved that with a static method in EditorInspector.

Note that the new changes are smaller in scope. I only handled EditorInspector, EditorInspectorCategory and EditorInspectorSection. I'll add cache to remaining inspector controls in a follow-up.

Needs re-reviewing.

akien-mga commented:
Thanks!
